### PR TITLE
Disable eval dataloaders replacement during overfitting

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Follow up: #9709

- Doc update
- Remove eval dataloader replacement logic
- add a test for a distributed sampler with overfit_batches
- also moved 2 tests to where they are should be (with some improvements)

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @justusschock @awaelchli @ninginthecloud

### CODE REVIEW

File: docs/source/common/debugging.rst [hunk 1]
```diff
@@ -79,10 +79,10 @@ argument of :class:`~pytorch_lightning.trainer.trainer.Trainer`)
 
 .. testcode::
 
-    # use only 1% of training data (and use the same training dataloader (with shuffle off) in val and test)
+    # use only 1% of training data (and turn-off validation)
```
Comment (awaelchli) [inline docs/source/common/debugging.rst:82]:
```suggestion
    # use only 1% of training data (and turn off validation)
```

File: docs/source/common/trainer.rst [hunk 2]
```diff
@@ -1074,7 +1074,7 @@ overfit_batches
 
 |
 
-Uses this much data of the training set. If nonzero, will use the same training set for validation and testing.
+Uses this much data of the training set. If nonzero, will turn-off validation.
```
Comment (awaelchli) [inline docs/source/common/trainer.rst:1077]:
```suggestion
Uses this much data of the training set. If nonzero, will turn off validation.
```

File: docs/source/common/debugging.rst (hunk 1 above)
Comment (awaelchli) [reply docs/source/common/debugging.rst:82]: "turn-off" is a noun if I understand correctly

File: docs/source/guides/speed.rst [hunk 3]
```diff
@@ -336,7 +336,7 @@ If you don't want to check 100% of the training/validation/test set set these fl
 
 If you also pass ``shuffle=True`` to the dataloader, a different random subset of your dataset will be used for each epoch; otherwise the same subset will be used for all epochs.
 
-.. note:: ``limit_train_batches``, ``limit_val_batches`` and ``limit_test_batches`` will be overwritten by ``overfit_batches`` if ``overfit_batches`` > 0. ``limit_val_batches`` will be ignored if ``fast_dev_run=True``.
+.. note:: ``limit_train_batches`` will be overwritten by ``overfit_batches`` if ``overfit_batches > 0`` and will turn-off validation.
```
Comment (awaelchli) [inline docs/source/guides/speed.rst:339]:
```suggestion
.. note:: ``limit_train_batches`` will be overwritten by ``overfit_batches`` if ``overfit_batches > 0`` and will turn off validation.
```

File: tests/trainer/flags/test_limit_batches.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
+# Copyright The PyTorch Lightning team.
```
Comment (awaelchli) [inline tests/trainer/flags/test_limit_batches.py:1]: in this file and the other, from the set of tests you moved here, did you need to make any changes?

File: tests/trainer/test_data_loading.py [hunk 5]
```diff
@@ -335,3 +335,19 @@ def test_pre_made_batches():
...
+
+
+@RunIf(skip_windows=True)
+def test_distributed_sampler_with_overfit_batches(tmpdir):
+    model = BoringModel()
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        overfit_batches=1,
+        fast_dev_run=1,
+        strategy="ddp_find_unused_parameters_false",
```
Comment (awaelchli) [inline tests/trainer/test_data_loading.py:347]: a particular reason for this choice? "ddp_spawn" would normally suffice

Comment (kaushikb11) [reply tests/trainer/test_data_loading.py:347]: But `num_processes` is 1?

File: tests/trainer/flags/test_limit_batches.py (hunk 4 above)
Comment (rohitgr7) [reply tests/trainer/flags/test_limit_batches.py:1]: yes.. just replaced the EvalModel with BoringModel. Nothing major.

File: pytorch_lightning/trainer/data_loading.py [hunk 6]
```diff
@@ -293,28 +292,14 @@ def _reset_eval_dataloader(
...
-                    )
-                    dataloaders[loader_i] = _update_dataloader(loader, SequentialSampler(loader.dataset), mode=mode)
-                else:
-                    rank_zero_warn(
-                        f"Your `{mode.dataloader_prefix}_dataloader` has `shuffle=True`,"
-                        "it is strongly recommended that you turn this off for val/test/predict dataloaders."
-                    )
+                rank_zero_warn(
+                    f"Your `{mode.dataloader_prefix}_dataloader` has `shuffle=True`,"
+                    "it is strongly recommended that you turn this off for val/test/predict dataloaders."
```
Comment (carmocca) [inline pytorch_lightning/trainer/data_loading.py:301]: Use `PossibleUserWarning`? #10675

File: tests/trainer/test_data_loading.py (hunk 5 above)
Comment (carmocca) [reply tests/trainer/test_data_loading.py:347]: This was the old mechanism to force "DDP" on "CPU"

nowadays just `strategy="ddp"` works on CPU. I guess this was copied from an old test

Comment (rohitgr7) [reply tests/trainer/test_data_loading.py:347]: yeah.. looks like the case.. we don't update the provided strategy in case num_processes stays 1

Comment (rohitgr7) [reply tests/trainer/test_data_loading.py:347]: it's actually a bug here I think. when you pass just ddp or ddp_spawn, it defaults to single device plugin but when provided with registry arguments it doesn't.

Comment (rohitgr7) [reply tests/trainer/test_data_loading.py:347]: I don't think so.. with `ddp` it goes on single device plugin but with `ddp_something_false` it stays in ddp

**Review by carmocca — APPROVED** [review]

File: tests/trainer/test_data_loading.py [hunk 7]
```diff
@@ -335,3 +335,19 @@ def test_pre_made_batches():
...
+
+
+@RunIf(skip_windows=True)
+def test_distributed_sampler_with_overfit_batches(tmpdir):
+    model = BoringModel()
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        overfit_batches=1,
+        fast_dev_run=1,
+        strategy="ddp",
```
Comment (rohitgr7) [inline tests/trainer/test_data_loading.py:347]:
```suggestion
        strategy="ddp_spawn",
```

**Review by awaelchli — APPROVED** [review]

File: tests/trainer/flags/test_limit_batches.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+        (RunningStage.PREDICTING, "predict"),
+    ],
+)
+def test_eval_limit_batches(stage, mode):
+    limit_eval_batches = f"limit_{mode}_batches"
+    dl_hook = f"{mode}_dataloader"
+    model = BoringModel()
+    eval_loader = getattr(model, dl_hook)()
+
+    limit_batches = 0.1
```
Comment (Borda) [inline tests/trainer/flags/test_limit_batches.py:61]: can we rather use this as a parameter, so it will be later easier to separate the failing cases...

File: tests/trainer/flags/test_overfit_batches.py [hunk 9]
```diff
@@ -62,3 +64,104 @@ def train_dataloader(self):
...
+
+    # ------------------------------------------------------
+    # get the training loader and batch
+    # ------------------------------------------------------
+    # Create a reference train dataloader without shuffling.
+    train_loader = DataLoader(dm.train_dataloader().dataset, shuffle=False)
+    (xa, ya) = next(iter(train_loader))
+    train_loader = DataLoader(dm.train_dataloader().dataset, shuffle=True)
+    full_train_samples = len(train_loader)
+    num_train_samples = int(0.11 * full_train_samples)
```
Comment (Borda) [inline tests/trainer/flags/test_overfit_batches.py:92]: any special reason for 0.11? :rabbit:

File: tests/trainer/flags/test_overfit_batches.py [hunk 10]
```diff
@@ -62,3 +64,104 @@ def train_dataloader(self):
...
+            # make sure we turned off shuffle for the user
+            assert isinstance(dataloaders[0].sampler, SequentialSampler)
+
+        trainer = Trainer(overfit_batches=5)
+        trainer._data_connector.attach_dataloaders(model)
+        loader_num_batches, _ = trainer._reset_eval_dataloader(split, model=model)
+        if split == RunningStage.VALIDATING:
+            assert loader_num_batches[0] == 0
+        else:
+            assert loader_num_batches[0] == len(test_loader)
```
Comment (Borda) [inline tests/trainer/flags/test_overfit_batches.py:167]: can we have this as a separate test with two parameters: `split` and `overfit_batches`

File: tests/trainer/flags/test_overfit_batches.py [hunk 11]
```diff
@@ -62,3 +64,104 @@ def train_dataloader(self):
...
+
+    # set the model loaders
+    model.train_dataloader = lambda: train_loader
+    model.val_dataloader = lambda: val_loader
+    model.test_dataloader = lambda: test_loader
+
+    # ------------------------------------------------------
+    # test train loader applies correct limits
+    # ------------------------------------------------------
+    trainer = Trainer(overfit_batches=4)
```
Comment (Borda) [inline tests/trainer/flags/test_overfit_batches.py:108]: I would split it here and use `overfit_batches` as parameter

File: tests/trainer/test_trainer_tricks.py [hunk 12]
```diff
@@ -1,174 +0,0 @@
-# Copyright The PyTorch Lightning team.
```
Comment (Borda) [inline tests/trainer/test_trainer_tricks.py:1]: ok, I see how it came, but pls lets improve these tests :]

Comment (rohitgr7) [reply tests/trainer/test_trainer_tricks.py:1]: done!

File: tests/trainer/flags/test_limit_batches.py (hunk 8 above)
Comment (rohitgr7) [reply tests/trainer/flags/test_limit_batches.py]: done!

**Review by Borda — APPROVED** [review]
